ai_agents/agents/ten_packages/extension/bytedance_llm_based_asr/extension.py

```diff
@@ -383,39 +447,96 @@ async def _on_asr_result(self, result: ASRResponse) -> None:
-            if is_final:
-                await self._finalize_end()
-
-            actual_start_ms = int(
-                self.audio_timeline.get_audio_duration_before_time(
+            # Process utterances: send definite=true individually,
+            # and concatenate adjacent definite=false utterances together
+            if not result.utterances:
+                # No utterances, send result.text as fallback (non-final)
+                # Use result.start_ms for fallback case
+                actual_start_ms = self._calculate_utterance_start_ms(
                     result.start_ms
                 )
-                + self.sent_user_audio_duration_ms_before_last_reset
-            )
-
-            asr_result = ASRResult(
-                id=str(uuid.uuid4()),  # Generate unique ID for each result
-                text=result.text,
-                final=is_final,
-                start_ms=actual_start_ms,
-                duration_ms=result.duration_ms,
-                language=result.language,
-                words=[],  # Empty list instead of None
-            )
+                await self._send_asr_result_from_text(
+                    text=result.text,
+                    is_final=False,
+                    start_ms=actual_start_ms,
+                    duration_ms=result.duration_ms,
+                    language=result.language,
+                    metadata={},
+                )
+                return
 
-            # Send result to TEN environment
-            await self.send_asr_result(asr_result)
+            # Group utterances: process each definite=true individually,
+            # and concatenate adjacent definite=false utterances
+            has_final_result = False
+            for is_definite, group in itertools.groupby(
+                result.utterances, key=lambda utt: utt.definite
+            ):
+                utterances_in_group = list(group)
+
+                if is_definite:
+                    # definite=true: send each individually with its own metadata
+                    has_final_result = True
+                    for utterance in utterances_in_group:
+                        # Calculate start_ms and duration_ms for this utterance
+                        actual_start_ms = self._calculate_utterance_start_ms(
+                            utterance.start_time
+                        )
+                        duration_ms = utterance.end_time - utterance.start_time
+
+                        metadata = self._extract_metadata(utterance)
+                        await self._send_asr_result_from_text(
+                            text=utterance.text,
+                            is_final=True,
+                            start_ms=actual_start_ms,
+                            duration_ms=duration_ms,
+                            language=result.language,
+                            metadata=metadata,
+                        )
+                else:
```

Comment:
There shouldn't be any difference between these two approaches.  Either send each utterance individually, or concatenate them based on whether they are the same definite utterance before sending them. There shouldn't be any special logic involved.